xls/modules/rle/rle.x

```diff
@@ -0,0 +1,324 @@
+    prev: TestEncOutData, // previous data received from the RLE encoder
+}
+
+// auxiliary proc for sending data to the tested RLE encoder
+proc RLEEncTestSender {
+    enc_input_s: chan<TestEncInData> out;
+
+    init {(u32:0)}
+
+    config (
+        enc_input_s: chan<TestEncInData> out,
+    ) {(enc_input_s,)}
+
+    next (tok: token, counter: u32) {
+        if counter < array_size(TEST_SEND_LOOKUP) {
+            let data = TEST_SEND_LOOKUP[counter];
+            let send_tok = send(tok, enc_input_s, data);
+            let _ = trace_fmt!("Sent {} transactions, symbol: 0x{:x}, last: {}", counter, data.symbol, data.last);
+            counter + u32:1
+         } else {
+            counter
+         }
+    }
+}
+
+// main proc used to test the RLE encoder
+#[test_proc]
+proc RLEEncTester {
+    terminator: chan<bool> out;            // test termination request
+    enc_output_r: chan<TestEncOutData> in; // data read from the tested RLE encoder
+
+    init {(
+        RLEEncTesterState {
+            recv_count: u32:0,
+            total_count: u32:0,
+            prev: TestEncOutData  {
+                symbol: TestSymbol:0,
+                count: TestCount:0,
+                last: false,
+            },
+        }
+    )}
+
+    config(terminator: chan<bool> out) {
+        let (enc_input_s, enc_input_r) = chan<TestEncInData>;
+        let (enc_output_s, enc_output_r) = chan<TestEncOutData>;
+
+        spawn RLEEncTestSender(enc_input_s);
+        spawn RLEEnc<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>(enc_input_r, enc_output_s);
+        (terminator, enc_output_r)
+    }
+
+    next(tok: token, state: RLEEncTesterState) {
+        let (recv_tok, enc_output) = recv(tok, enc_output_r);
+
+        let recv_count = state.recv_count + u32:1;
+        let total_count = state.total_count + enc_output.count as u32;
+
+        let _ = trace_fmt!(
+            "Received {} transactions, symbol: 0x{:x}, count: {}, last: {}, total of {} symbols",
+            recv_count, enc_output.symbol, enc_output.count, enc_output.last, total_count
+        );
+
+        // checks for expected values
+
+        let exp_enc_output = TEST_RECV_LOOKUP[state.recv_count];
+        let _ = assert_eq(enc_output, exp_enc_output);
+
+        // check for general properties of the received data
+
+        let max_count = std::unsigned_max_value<TEST_COUNT_WIDTH>();
+
+        let _ = assert_eq(enc_output.count > TestCount:0, true);
+        let _ = assert_eq(enc_output.count <= max_count, true);
+
+        // if the symbol repeats, check that the symbol counter has reached
+        // its maximum value or if the symbol was last in the transaction
+
+        let prev_data_valid = state.recv_count != u32:0;
+        let _ = if prev_data_valid {
+            let data_repeats = state.prev.symbol == enc_output.symbol;
+            let overflow = state.prev.count == max_count;
+            let last = state.prev.last;
+
+            let _ = if data_repeats {
```

Comment:
wouldn't `assert_eq(data-repeats, overflow || last)` do the same?

Reply:
Not exactly, as we want to skip assert entirely if data_didn't repeat.
It can be changed to `assert_eq(data_repeats && (overflow || last), data_repeats)`,
but I'm not sure if this will make this code more readable.